man/sarp.8
.\"  sarp -- client for Secure ARP key management
.\"
.\"  Copyright (C) 2002  ALoR <alor@users.sourceforge.net>
.\"
.\"  This program is free software; you can redistribute it and/or modify
.\"  it under the terms of the GNU General Public License as published by
.\"  the Free Software Foundation; either version 2 of the License, or
.\"  (at your option) any later version.
.\"
.\"  This program is distributed in the hope that it will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"
.\"  You should have received a copy of the GNU General Public License
.\"  along with this program; if not, write to the Free Software
.\"  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\"
.de Sp
.if n .sp
.if t .sp 0.4
..
.TH SARP "8" "@SERIAL@" "sarp 0.0.9-devel"
.SH NAME
.B sarp 0.0.9-devel \- client for Secure ARP key management
.SH SYNOPSIS
.B sarp
[\fIOPTIONS\fR] 

.SH DESCRIPTION
sarp is a tool for the creation of DSA key to be used in a S-ARP LAN.
.br
the keys have to be put in the AKD database. 

.PP
.SH OPTIONS
Options that make sense together can generally be combined. sarpd will warn the user
about unsupported option combinations.
.TP
\fB\-b\fR, \fB\-\-bitlen <FILE>\fR
use this option to generate keys with arbitrary key length (def. 1024)
.TP
\fB\-g\fR, \fB\-\-genkey\fR
generate a keypair
.TP
\fB\-s\fR, \fB\-\-sign\fR
generate a signed info file. this file format is used by the AKD to load the
information in its database. 
.TP
\fB\-o\fR, \fB\-\-outfile <FILE>\fR
specify the output filename for keypair
.TP
\fB\-c\fR, \fB\-\-check\fR
check if a given keypair is valid. it perform a test to check if the public
key corresponts to the private one
.TP
\fB\-i\fR, \fB\-\-infile <FILE>\fR
specify the input filename for keypair verification
.TP
\fB\-v\fR, \fB\-\-verbose\fR
be verbose during operations
.TP
\fB\-V\fR, \fB\-\-version\fR
prints the version and exit

                  
.SH EXAMPLES
Here are some examples of using sarpd.
.TP
.B sarp -gv -b 512
.Sp
create a 512 bit key pair
.TP
.B sarp -gsv
.Sp
create a 1024 bit signed key pair (the .sarp file is the file to be put in the
AKD repository directory)

.SH PLATFORMS
Linux 2.2.x  2.4.x


.SH AUTHOR
Alberto Ornaghi (ALoR) <alor@security.dico.unimi.it>





CHANGELOG

Legenda:
         + new feature
         - old feature removed
         ! bug fixed

=========================================

0.0.9 -- 2002

+ sarp header now include timestamp to avoid replay attack
+ hosts are syncronized with the CA on startup
+ if the CA(s) are not respondig the daemon will shutdown
! fixed the time calculation for timing functions
! drop malformed packets


0.0.8 -- 20021007

+ sarp can now create keypair of different bit lenght
+ use DSA_sign_setup to precalculate DSA prameter (kinv and r)
+ LKM ported to kernel 2.2.x
! fixed a bug in the CA: loading ip addresses not belonging to any iface
! fixed : accept a reply KEY_NOTFOUND not signed

   basic setup (1024 bit keypair):
      + 3 host (2 peons and 1 CA)
      + ping from peon to peon (key not in cache) time=49.6 ms
      + ping from peon to peon (key in cache)     time=12.8 ms
   basic setup (512 bit keypair):
      + 3 host (2 peons and 1 CA)
      + ping from peon to peon (key not in cache) time=19.5 ms
      + ping from peon to peon (key in cache)     time=5.3 ms

 
0.0.7 -- 20021001

+ the CA is now complete
! fixed many memory leak
! fixed a bug in the LKM returning 0 if sarp was already loaded

   basic setup (1024 bit keypair):
      + 3 host (2 peons and 1 CA)
      + ping from peon to peon (key not in cache) time=220.3 ms
      + ping from peon to peon (key in cache)     time=22.4 ms
   basic setup (512 bit keypair):
      + 3 host (2 peons and 1 CA)
      + ping from peon to peon (key not in cache) time=34.1 ms
      + ping from peon to peon (key in cache)     time=8.8 ms


0.0.6 -- 20020925

+ added the 'sarp' command
   + keys generation
   + keys validation
+ dsa sign and verify
+ configuration file
+ CA list (one CA per interface)
+ store and retrive an host key


0.0.5 -- 20020914

+ the first REAL run was successful ;)
  (i've tried it without signatures, but it works! it correctly disable the 
   kernel, take the arp reply, look in known_host and add it to the neigh via
   socket netlink)
+ all the infrastructure is ready (now its time for criptography ;) )
   + neighbor management
   + sarp parsing
   + compatibility mode with arp reply


0.0.4 -- 20020903

+ libnetlink support (neighbor manipulation)
+ new configure for libs
! fixed a bug in the search_known_host logic


0.0.3 -- 20020901

+ libnet support (send_to_wire)
+ known_hosts hash table
+ preliminar ARP parsing (with reply)
+ --prefix to configuration files


0.0.2 -- 20020831

+ version 0.8 for sarp.o (lkm)
+ sarpd now listens on "any" iface


0.0.1 -- 20020829

+ moved the /proc/net/sarp entry in the more appropriate
   /proc/sys/net/ipv4/sarp location
+ kernel patch ready
+ LKM code ready (without the automake file)


0.0.0 -- 20020826

* Here we go...

+ Makefiles and configure ready (with autoconf and automake)
+ directory structure
+ spec for rpm packaging
+ initial man page



kernel/module/sarp.c
/*
    sarp -- LKM for Secure ARP authentication

    Copyright (C) 2002  ALoR <alor@users.sourceforge.net>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.

   history:

   0.9 - ported to kernel 2.2.x (thanks to gigi sullivan)
   0.8 - added CONFIG_MODVERSIONS (thanks to gigi sullivan)
   0.7 - arp_packet_type address is from System.map (passed by Makefile)
   0.5 - disable secure arp on exit
   0.3 - moved in /proc/sys/net/ipv4/sarp
   0.1 - inizial version - extracted fro kernel patch
   
   0.X - make it work on kernel 3.2.0
	 Oct 29, 2013
	 daveti@cs.uoregon.edu
	 http://davejingtian.org
 
*/

//daveti: remove these below
//#define __KERNEL__
//#define MODULE

//daveti: kernel version checking for autoconf.h
#include <linux/version.h>
#if LINUX_VERSION_CODE < KERNEL_VERSION(2, 6, 33)
#include <linux/autoconf.h>
#else
#include <generated/autoconf.h>
#endif
//#include <linux/autoconf.h>
//daveti: remove modversions.h and config.h
//#ifdef CONFIG_MODVERSIONS
//#include <linux/modversions.h>
//#endif
//#include <linux/config.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/kernel.h>
#include <linux/sched.h>
#include <linux/proc_fs.h>
#include <linux/version.h>
#include <asm/uaccess.h>

#include <linux/netdevice.h>  /* dev_[add|remove]_pack */
#include <net/arp.h>          /* arp_recv */


#define MODULE_NAME    "sarp"
//daveti: rename module_version
//#define MODULE_VERSION "0.9"
#define SARP_VERSION "1.0"
//daveti: unable to create a proc entry under /proc/sys/net/ipv4 using create_proc_entry...
#define SARP_PROC_PATH "sarp"

MODULE_AUTHOR("Alberto Ornaghi");
MODULE_DESCRIPTION("Secure ARP");
#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,0)
MODULE_LICENSE("GPL");
#endif

/* 
 * this structure is declared in net/ipv4/arp.c 
 * we need its address.
 * so Makefile greps the /boot/System.map searching for it
 * then it pass the value in the ARP_PACKET_TYPE_ADDR
 */

struct packet_type *arp_packet_type = (void *)ARP_PACKET_TYPE_ADDR;

#if 0
static struct packet_type arp_packet_type = {
   type: __constant_htons(ETH_P_ARP),
   func: arp_rcv,
   data: (void*) 1, /* understand shared skbs */
};
#endif

/* globals */

int secure_arp_enabled = 0;

/* protos */

int secure_arp_proc_read (char *buf, char **start, off_t offs, int len);
ssize_t secure_arp_proc_write( struct file *file, const char *buf, size_t length, loff_t *offset);
void enable_sarp(void);
void disable_sarp(void);


/*************************************/

void enable_sarp(void)
{
   secure_arp_enabled = 1;
   
   dev_remove_pack(arp_packet_type);
   
   printk(KERN_INFO "[secure_arp] enabled\n");
   printk(KERN_INFO "[secure_arp] kernel can now reveive ARP entries "\
                          "only through SARP daemon\n");
}


void disable_sarp(void)
{
   secure_arp_enabled = 0;
         
   dev_add_pack(arp_packet_type);
         
   printk(KERN_INFO "[secure_arp] disabled\n");
   printk(KERN_INFO "[secure_arp] kernel can now receive \"classic\" "\
                          "ARP packets\n");
}


int secure_arp_proc_read (char *buf, char **start, off_t offs, int len) 
{  
   int written;
   
   //MOD_INC_USE_COUNT;
   
   written = sprintf(buf, "%d\n", secure_arp_enabled);

   //MOD_DEC_USE_COUNT;
   
   return written;
}

ssize_t secure_arp_proc_write( struct file *file, const char *buf, size_t length, loff_t *offset)
{
   #define MESSAGE_LEN 5
   int i, value;
   char *message;

   //MOD_INC_USE_COUNT;

   message = kmalloc(MESSAGE_LEN, GFP_KERNEL);

   for (i = 0; i < MESSAGE_LEN-1 && i < length; i++)
      get_user(message[i], buf + i);
   
   message[i]='\0';
   value = simple_strtoul(message, NULL, 10);
   kfree(message);
 
   switch(value) {
      case 1:   /* enable it */
         if (secure_arp_enabled) {
            //MOD_DEC_USE_COUNT;
            return i;
         }
         
         enable_sarp();      
         
         break;
      case 0:   /* disable it */
         if (!secure_arp_enabled) {
            //MOD_DEC_USE_COUNT;
            return i;
         }
   
         disable_sarp();
         
         break;
      default:  /* error */
         //MOD_DEC_USE_COUNT;
         return -1;
         break;
   }
  
   //MOD_DEC_USE_COUNT;
  
   return i;                                                
}

/*
 * init and exit functions
 */


static int __init sarp_init(void)
{
   struct proc_dir_entry *sarp_entry;
  
//daveti: update the proc entry 
   //sarp_entry = create_proc_entry("sys/net/ipv4/sarp", 0644, NULL);
   sarp_entry = create_proc_entry(SARP_PROC_PATH, 0644, NULL);
  
//daveti: remove owner 
//#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,0)
//  sarp_entry->owner = THIS_MODULE;
//#endif
   sarp_entry->read_proc = (read_proc_t *)&secure_arp_proc_read;
   sarp_entry->write_proc = (write_proc_t *)&secure_arp_proc_write;
  
//daveti: update 
   //printk(KERN_INFO "%s %s module loaded\n", MODULE_NAME, MODULE_VERSION);
   printk(KERN_INFO "%s %s module loaded\n", MODULE_NAME, SARP_VERSION);
   return 0;
}


static void __exit sarp_exit(void)
{
//daveti: update proc entry
   //remove_proc_entry("sys/net/ipv4/sarp", NULL);
   remove_proc_entry(SARP_PROC_PATH, NULL);

   if (secure_arp_enabled)
      disable_sarp();
  
//daveti: update 
   //printk(KERN_INFO "%s %s removed\n", MODULE_NAME, MODULE_VERSION);
   printk(KERN_INFO "%s %s removed\n", MODULE_NAME, SARP_VERSION);
}

//daveti: update
//EXPORT_NO_SYMBOLS;
//#if LINUX_VERSION_CODE >= KERNEL_VERSION(2,4,0)
module_init(sarp_init);
module_exit(sarp_exit);
//#elif LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0)
/*
void cleanup_module(void)
{
   sarp_exit();
}

int init_module(void)
{
   return sarp_init();
}
#endif
*/

/* EOF */

// vim:ts=3:expandtab

